k8s/src/main/scala/io/buoyant/k8s/resources.scala

```diff
@@ -226,19 +224,14 @@ private[k8s] class NsObjectResource[O <: KubeObject: TypeReference, W <: Watch[O
     fieldSelector: Option[String] = None
   ): Future[(Seq[W], Option[String])] =
     get(
-      labelSelector,
-      fieldSelector,
-      None,
-      retryIndefinitely = true,
-      watch = true
+      labelSelector = labelSelector,
+      fieldSelector = fieldSelector,
+      retryIndefinitely = true
     ).map { maybeObj =>
       val watch = maybeObj.toSeq.flatMap { obj => Seq(od.toWatch(obj)) }
-      val version =
-        for {
-          obj <- maybeObj
-          meta <- obj.metadata
-          version <- meta.resourceVersion
-        } yield version
-      (watch, version)
+      // object watches don't supply a resource version, since the version associated
```

Comment:
I don't really understand.

Reply:
I can't seem to find this documented anywhere, but as far as I can tell, Kubernetes list resources (e.g. `/api/v1/namespaces/ns/services`) return the global resource version as part of the metadata response. This means that you can always use the version returned to establish a new watch.

On the other hand, Kubernetes object resources (e.g. `/api/v1/namespaces/ns/services/mysvc`) do not return the global resource version. They instead return the resource version from when the object was last modified, and that version might be too old to establish a new watch.

For instance, testing in one of our clusters, I make the following request for an endpoints object:

```
# curl localhost:8001/api/v1/namespaces/test/endpoints/namerd
{
  "kind": "Endpoints",
  "apiVersion": "v1",
  "metadata": {
    "name": "namerd",
    "namespace": "test",
    "selfLink": "/api/v1/namespaces/test/endpoints/namerd",
    "uid": "28d76695-af7d-11e7-a88a-42010af00004",
    "resourceVersion": "65078108",
    "creationTimestamp": "2017-10-12T18:42:54Z"
  },
```

That "resourceVersion" reflects when the endpoints were last modified, rather than the global resource version. If I then immediately make a watch request with the returned resource version, I get:

```
# curl localhost:8001/api/v1/watch/namespaces/test/endpoints/namerd?resourceVersion=65078108
{"type":"ERROR","object":{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"too old resource version: 65078108 (65950030)","reason":"Gone","code":410}}
```

So when watching objects, we need to omit the version, rather than risk having the watch fail because the resource version is too old.

Would love to get a bit more clarity on this though from folks with more familiarity of Kubernetes internals.

Reply:
Have confirmed that this is the same behavior in Kubernetes 1.7 as well.

Reply:
@klingerf if the resource version is actually "65078108" and we are seeing above error, that to me sounds like a k8s bug. Based on my experience working with zk/consul, the client should always send a last known version because the client could miss the events when the client is not watching. So if the client has an old version, k8s should return what has changed and latest resource version. I hope we tested that linkerd doesn't miss a watch change b/w the retries. But again, i'm speaking based on my past experience and may not be true for k8s.